messaging/coap/engine.c

```diff
@@ -168,775 +170,906 @@ coap_audit_log(const oc_message_t *msg)
-        }
-
-        oc_client_cb_invoke(message, &response_buffer, client_cb,
-                            &msg->endpoint);
-        /* Do not free the response buffer in case of a separate response
-         * signal from the server. In this case, the client_cb continues
-         * to live until the response arrives (or it times out).
-         */
-        if (oc_ri_is_client_cb_valid(client_cb)) {
-          if (client_cb->separate == 0) {
-            if (response_buffer) {
-              response_buffer->ref_count = 0;
-            }
-          } else {
-            client_cb->separate = 0;
-          }
-        }
-        goto send_message;
-#else  /* OC_BLOCK_WISE */
-        oc_client_cb_invoke(message, client_cb, &msg->endpoint);
-#endif /* OC_BLOCK_WISE */
-      }
+  if (client_cb != NULL) {
+    return coap_invoke_client_cb(ctx, endpoint, client_cb);
+  }
 #endif /* OC_CLIENT */
-    }
-  } else {
+  return COAP_SEND_EMPTY_MESSAGE;
+}
+
+int
+coap_receive(coap_receive_ctx_t *ctx, oc_endpoint_t *endpoint,
+             coap_make_response_fn_t response_fn, void *response_fn_data)
+{
+  /* handle requests */
+  if (ctx->request->code >= COAP_GET && ctx->request->code <= COAP_DELETE) {
+    return coap_receive_method(ctx, endpoint, response_fn, response_fn_data);
+  }
+  return coap_receive_code(ctx, endpoint);
+}
+
+int
+coap_process_inbound_message(oc_message_t *msg)
+{
+  OC_DBG("CoAP Engine: received datalen=%u from", (unsigned int)msg->length);
+  OC_LOGipaddr(msg->endpoint);
+  OC_LOGbytes(msg->data, msg->length);
+
+  /* static declarations reduce stack peaks and program code size */
+  static coap_packet_t message;
+  static coap_packet_t response;
+  static coap_receive_ctx_t ctx;
+
+  coap_status_t status;
+#ifdef OC_TCP
+  if ((msg->endpoint.flags & TCP) != 0) {
+    status = coap_tcp_parse_message(&message, msg->data, msg->length, false);
+  } else
+#endif /* OC_TCP */
+  {
+    status = coap_udp_parse_message(&message, msg->data, msg->length, false);
+  }
+
+  coap_set_global_status_code(status);
+
+  if (status != COAP_NO_ERROR) {
     OC_ERR("Unexpected CoAP command");
 #ifdef OC_SECURITY
     coap_audit_log(msg);
 #endif /* OC_SECURITY */
-    coap_status_t cs = coap_global_status_code();
-    if (msg->endpoint.flags & TCP) {
-      coap_send_empty_response(COAP_TYPE_NON, 0, message->token,
-                               message->token_len, (uint8_t)cs, &msg->endpoint);
-    } else {
-      coap_send_empty_response(message->type == COAP_TYPE_CON ? COAP_TYPE_ACK
-                                                              : COAP_TYPE_NON,
-                               message->mid, message->token, message->token_len,
-                               (uint8_t)cs, &msg->endpoint);
+#ifdef OC_TCP
+    if ((msg->endpoint.flags & TCP) != 0) {
+      coap_send_empty_response(COAP_TYPE_NON, 0, message.token,
+                               message.token_len, (uint8_t)status,
+                               &msg->endpoint);
```

Comment:
pls remove - this is a but. You cannot send empty message for TCP because it is defined only for UDP

Reply:
changed to `oc_connectivity_end_session`